File: 200. Number of Islands/main.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rowSize = len(grid)
        colSize = len(grid[0])

        def makeIslandsZero(grid: List[List[str]], i: int, j: int):
            if i < 0 or i >= rowSize or j < 0 or j >= colSize or grid[i][j] == "0":
                return

            grid[i][j] = "0"

            makeIslandsZero(grid, i - 1, j)
            makeIslandsZero(grid, i + 1, j)
            makeIslandsZero(grid, i, j - 1)
            makeIslandsZero(grid, i, j + 1)

        count = 0

        for i in range(rowSize):
            for j in range(colSize):
                if grid[i][j] == "1":
                    count += 1
                    makeIslandsZero(grid, i, j)

        return count

File: 200. Number of Islands/main.rs
struct Solution;

impl Solution {
    fn make_islands_zero(grid: &mut Vec<Vec<char>>, i: i32, j: i32) {
        let row_size = grid.len();
        let col_size = grid[0].len();

        if i < 0
            || i >= row_size as i32
            || j < 0
            || j >= col_size as i32
            || grid[i as usize][j as usize] == '0'
        {
            return;
        }

        grid[i as usize][j as usize] = '0';

        Self::make_islands_zero(grid, i - 1, j);
        Self::make_islands_zero(grid, i + 1, j);
        Self::make_islands_zero(grid, i, j - 1);
        Self::make_islands_zero(grid, i, j + 1);
    }

    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let row_size = grid.len();
        let col_size = grid[0].len();
        let mut count = 0;
        let mut grid = grid;

        for i in 0..row_size {
            for j in 0..col_size {
                if grid[i][j] == '1' {
                    count += 1;
                    Self::make_islands_zero(&mut grid, i as i32, j as i32);
                }
            }
        }

        count
    }
}
